File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        queue=[root]
        swap=True
        r=[]
        while queue:
            levelsize=len(queue)
            levelnode=[]
            for _ in range(levelsize):
                temp=queue.pop(0)
                levelnode.append(temp.val)
                if temp.left:
                    queue.append(temp.left)
                if temp.right:
                    queue.append(temp.right)
            if not swap:
                r.append(levelnode[::-1])
                swap=not swap
            else:
                r.append(levelnode)
                swap=not swap
        return r
        

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        List<List<Integer>> l=new ArrayList<>();
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        boolean flag=true;
        while(!q.isEmpty()){
            int size=q.size();
            List<Integer> g=new LinkedList<>();
            for(int i=0;i<size;i++){
                TreeNode node=q.poll();
                    g.add(node.val);
                
                if(node.left!=null){
                    q.add(node.left);
                }if(node.right!=null){
                    q.add(node.right);
                }
            }
            if(!flag){
                Collections.reverse(g);
            }
            flag=!flag;
            l.add(g);
        }
        return l;
    }
}